packages/next/src/server/base-server.ts

```diff
@@ -1768,6 +1813,8 @@ export default abstract class Server<ServerOptions extends Options = Options> {
 
     let resolvedUrlPathname = getRequestMeta(req, 'rewroteURL') || urlPathname
 
+    this.setVaryHeader(req, res, isAppPath, resolvedUrlPathname)
```

Comment:
As we don't do anything if the request isn't for an app path, maybe something like this instead:

```suggestion
    if (isAppPath) {
      this.setVaryHeader(req, res, resolvedUrlPathname)
    }
```

Noting as well that we don't need to delete the `Next-URL` header from non-app path requests as the pages router doesn't set it.

Reply:
See https://github.com/vercel/next.js/pull/61794#discussion_r1484523177